### DISCUSSION TITLE
Implementing Combined Authentication and Authorization for Web App and API

### DISCUSSION THREAD

**Discussion by Markus-Isaksson** [zitadel/zitadel#7441] [Q&A]
Hello Zitadel Community,

I'm currently working on a project that involves both a web application and an API running in the same process. The web application (a mix of server-side rendering and SPA) authenticates users via Zitadel's PKCE flow, and both the web app and other clients will make calls to the API. I'm relatively new to the concepts of authentication and authorization, OIDC, OAuth 2.0, etc., and I'm seeking guidance on how to properly secure the API with both authentication and authorization to support access by the web app as well as other external clients.

Here's a snippet of my current setup for the web app authentication using the Gorilla Mux router and Zitadel's OIDC for the PKCE flow. It's based on the [frontend documentation example for Go](https://zitadel.com/docs/examples/login/go):

```go
package main

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"github.com/gorilla/mux"
	"github.com/zitadel/zitadel-go/v3/pkg/authentication"
	openid "github.com/zitadel/zitadel-go/v3/pkg/authentication/oidc"
	"github.com/zitadel/zitadel-go/v3/pkg/zitadel"
)

func serverMain(conf ServerConfig) {
	ctx := context.Background()
	redirectURI := "http://localhost:8080/auth/callback" // Redirect URI registered at ZITADEL
	rolesClaim := fmt.Sprintf("urn:zitadel:iam:org:project:%s:roles", conf.Zitadel.ProjectId)

	authN, err := authentication.New(ctx, zitadel.New(conf.Zitadel.Issuer), conf.Zitadel.AuthSecret,
		openid.DefaultAuthentication(
			conf.Zitadel.ClientId,
			redirectURI,
			conf.Zitadel.AuthSecret,
			"openid", "profile", "email",
			rolesClaim,
			//cannot get this to work
			"urn:zitadel:iam:user:resourceowner:id",
			//cannot get this to work
			"urn:zitadel:iam:user:resourceowner:name",
		))

	if err != nil {
		fmt.Printf("Zitadel SDK could not initialize: %s", err)
		os.Exit(1)
	}

	mw := authentication.Middleware(authN)

	router := mux.NewRouter()
	appRouter := router.Host("localhost").Subrouter()
	apiRouter := router.Host("api.localhost").Subrouter()
	apiRouter.HandleFunc("/test", apiHandler).Methods("POST")
	appRouter.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		http.Redirect(w, req, "/signedin/projects", http.StatusFound)
	}))

	//The zitadel docs say to register the authentication endpoints like this
	//router.Handle("/auth/", authN)
	// It will register the following handlers on it:
	// - /login (starts the authentication process to the Login UI)
	// - /callback (handles the redirect back from the Login UI)
	// - /logout (handles the logout process)
	//However this doesn't work with the gorilla muxer.
	//This is because in Go's muxer a handler will match any subroute, thus .Handle("/auth/") will match "/auth/login".
	//This is not true in gorilla, hence this prefixing approach must be used.
	appRouter.PathPrefix("/auth").Handler(authN)

	authenticatedRouter := appRouter.PathPrefix("/signedin/").Subrouter()
	authenticatedRouter.Use(mw.RequireAuthentication())
	// ... additional routing for the web app
}

func apiHandler(w http.ResponseWriter, r *http.Request) {
	json := `{
		"foo": "bar"
	}`
	w.Header().Set("Content-Type", "application/json")
	w.Write([]byte(json))
}
```

This setup successfully handles user authentication for the web application. However, I need to extend this to secure my API endpoints, ensuring proper authentication and authorization for access by the web app and external clients. Moreover, I'm planning to introduce multi-tenancy and a role-based access control system within the web app to manage access to different resources and operations based on the user's role and their organization.

Could anyone provide guidance or point me to examples where both web app and API authentication and authorization are implemented together? Any advice on best practices or common pitfalls to avoid would also be greatly appreciated.

Thank you in advance for your help!

**Comment by dakshitha** [comment]
This example should give you an idea about how to set up the applications and scopes. It's Python based but should give basic information on setting things up for a front-end app calling a back-end API- https://zitadel.com/docs/guides/solution-scenarios/frontend-calling-backend-API

Let me know if you have more questions.
